File: 62.unique-paths.js
/*
 * @lc app=leetcode id=62 lang=javascript
 *
 * [62] Unique Paths
 */

// @lc code=start
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function (m, n) {
  const dp = Array.from({ length: m }, () =>
    Array.from({ length: n }, () => 1)
  );
  for (let i = 1; i < m; ++i) {
    for (let j = 1; j < n; ++j) {
      dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
    }
  }
  return dp[m - 1][n - 1];
};
// @lc code=end

File: solutions/62.unique-paths.rb
#
# @lc app=leetcode id=62 lang=ruby
#
# [62] Unique Paths
#

# @lc code=start
# @param {Integer} m
# @param {Integer} n
# @return {Integer}
def unique_paths(m, n)
  return 1 if m==1 || n==1
  $map[[m,n]] ||= unique_paths(m-1, n) + unique_paths(m, n-1) # map赋值后能直接返回'值'
end

$map = {}
# @lc code=end

